File: backend/firestore/test/group_ticket_audit.test.mjs
/**
 * A11b: auditoría del ticket por el grupo.
 *
 * La foto y el desglose de un gasto compartido son EVIDENCIA, no un archivo
 * privado de quien lo subió: cualquier miembro del grupo tiene que poder
 * comprobar que la línea interpretada coincide con el papel. Estas pruebas
 * fijan las dos mitades del contrato: la lectura se abre al grupo y la
 * escritura NO se mueve ni un milímetro.
 *
 * Ejecutar desde la raíz del repo:
 *   firebase emulators:exec --only firestore --project demo-salda \
 *     "npm --prefix backend/firestore test"
 */
import { readFileSync } from 'node:fs';
import { after, before, beforeEach, describe, it } from 'node:test';
import {
  assertFails,
  assertSucceeds,
  initializeTestEnvironment,
} from '@firebase/rules-unit-testing';
import {
  collection,
  deleteDoc,
  deleteField,
  doc,
  getDoc,
  getDocs,
  serverTimestamp,
  setDoc,
  updateDoc,
} from 'firebase/firestore';

let env;

// Quién es quién. El grupo `gr1` lo preside JEFA; ALBA sube el ticket y
// JORGE es un miembro normal que no consume nada de él.
const JEFA = 'uid-jefa';
const ALBA = 'uid-alba';
const JORGE = 'uid-jorge';
const ADMIN = 'uid-admin'; // miembro de gr1 con role: admin
const AJENA = 'uid-ajena'; // miembro de OTRO grupo
const EXTERNO = 'uid-externo'; // cuenta verificada sin relación alguna
const PAREJA = 'uid-pareja'; // la otra mitad de la relación rel1

const SG = 'sessions/sg1'; // sesión contextual del grupo gr1
const T = `${SG}/accounts/a1/tickets/t1`;
const SR = 'sessions/sr1'; // sesión contextual de la RELACIÓN rel1
const SL = 'sessions/sl1'; // sesión LEGACY, sin contexto

const db = (uid) =>
  env
    .authenticatedContext(uid, {
      email: `${uid}@salda.test`,
      email_verified: true,
      firebase: { sign_in_provider: 'password' },
    })
    .firestore();

before(async () => {
  env = await initializeTestEnvironment({
    projectId: 'demo-salda',
    firestore: {
      rules: readFileSync(
        new URL('../firestore.rules', import.meta.url),
        'utf8',
      ),
    },
  });
});
after(async () => env?.cleanup());

beforeEach(async () => {
  await env.clearFirestore();
  await env.withSecurityRulesDisabled(async (ctx) => {
    const f = ctx.firestore();

    // Identidad pública: sin ella no hay canUseSocial().
    for (const uid of [JEFA, ALBA, JORGE, ADMIN, AJENA, EXTERNO, PAREJA]) {
      await setDoc(doc(f, `profiles/${uid}`), {
        displayName: uid, displayNameLower: uid, username: uid,
        createdAt: serverTimestamp(), updatedAt: serverTimestamp(),
        schemaVersion: 1,
      });
    }

    // Grupo con cuatro miembros.
    await setDoc(doc(f, 'spaces/gr1'), {
      name: 'Piso', ownerUid: JEFA, kind: 'group', status: 'active',
      createdAt: serverTimestamp(), updatedAt: serverTimestamp(),
      schemaVersion: 2,
    });
    for (const uid of [JEFA, ALBA, JORGE, ADMIN]) {
      await setDoc(doc(f, `spaces/gr1/members/${uid}`), {
        uid, joinedAt: serverTimestamp(),
        ...(uid === ADMIN ? { role: 'admin' } : {}),
      });
    }

    // Otro grupo, sin nada que ver con el primero.
    await setDoc(doc(f, 'spaces/gr2'), {
      name: 'Otro', ownerUid: AJENA, kind: 'group', status: 'active',
      createdAt: serverTimestamp(), updatedAt: serverTimestamp(),
      schemaVersion: 2,
    });
    await setDoc(doc(f, `spaces/gr2/members/${AJENA}`), {
      uid: AJENA, joinedAt: serverTimestamp(),
    });

    // Relación: su modelo NO cambia con A11b.
    await setDoc(doc(f, 'spaces/rel1'), {
      name: 'Jefa y pareja', ownerUid: JEFA, kind: 'relationship',
      relationshipUids: [JEFA, PAREJA].sort(), status: 'active',
      createdAt: serverTimestamp(), updatedAt: serverTimestamp(),
      schemaVersion: 2,
    });
    for (const uid of [JEFA, PAREJA]) {
      await setDoc(doc(f, `spaces/rel1/members/${uid}`), {
        uid, joinedAt: serverTimestamp(),
      });
    }

    // El ticket de Alba, nacido DENTRO del grupo (ADR-030).
    await setDoc(doc(f, SG), {
      ownerUid: ALBA, kind: 'multi', name: 'Compra', status: 'open',
      splitModeDefault: 'byItem', shareCode: 'SECRET-CODE-16CHARS',
      currency: 'EUR', contextModelVersion: 1, spaceId: 'gr1',
      computeVersion: 0, totals: { grandTotal: 550 }, balances: {},
    });
    await setDoc(doc(f, `${SG}/participants/p1`), {
      name: 'Alba', isOwner: true, order: 0, claimedByDevice: '',
    });
    // Jorge entra en el reparto con su UID reclamado, como cualquier miembro
    // del grupo que el anfitrión añade al gasto (ADR-030).
    await setDoc(doc(f, `${SG}/participants/p2`), {
      name: 'Jorge', isOwner: false, order: 1, claimedByDevice: JORGE,
    });
    await setDoc(doc(f, `${SG}/accounts/a1`), { name: 'Súper', order: 0 });
    await setDoc(doc(f, T), {
      kind: 'scanned', grandTotal: 550, paidByParticipantId: 'p1',
      imagePath: 'receipts/sg1/t1/original.jpg',
      merchant: { name: 'Familycash' },
      spaceId: 'gr1', contextModelVersion: 1,
    });
    // Jorge no consume NADA de este ticket: aun así debe poder auditarlo.
    await setDoc(doc(f, `${T}/lines/l1`), {
      name: 'Coca-Cola', totalPrice: 250, quantityMilli: 1000, order: 0,
      assignment: { type: 'units', schemaVersion: 2, units: { u0: { p1: true } } },
    });
    await setDoc(doc(f, `${T}/lines/l2`), {
      name: 'Patatas', totalPrice: 300, quantityMilli: 1000, order: 1,
      unitIds: ['u0'],
      assignment: { type: 'units', schemaVersion: 2, units: {} },
    });
    await setDoc(doc(f, `${SG}/settlements/st1`), {
      from: 'p2', to: 'p1', amount: 250, state: 'pending',
    });

    // Mismo montaje dentro de una RELACIÓN.
    await setDoc(doc(f, SR), {
      ownerUid: JEFA, kind: 'single', name: 'Cena', status: 'open',
      splitModeDefault: 'equal', shareCode: 'SECRET-CODE-16CHARS',
      currency: 'EUR', contextModelVersion: 1, spaceId: 'rel1',
      computeVersion: 0, totals: { grandTotal: 900 }, balances: {},
    });
    await setDoc(doc(f, `${SR}/accounts/a1`), { name: 'Cena', order: 0 });
    await setDoc(doc(f, `${SR}/accounts/a1/tickets/t1`), {
      kind: 'manual', grandTotal: 900, paidByParticipantId: 'p1',
      spaceId: 'rel1', contextModelVersion: 1,
    });
    await setDoc(doc(f, `${SR}/accounts/a1/tickets/t1/lines/l1`), {
      name: 'Menú', totalPrice: 900,
      assignment: { type: 'unassigned', participants: {} },
    });

    // Sesión LEGACY (sin contexto) cuyo ticket se vinculó a mano al grupo.
    await setDoc(doc(f, SL), {
      ownerUid: ALBA, kind: 'single', name: 'Antiguo', status: 'open',
      splitModeDefault: 'equal', shareCode: 'SECRET-CODE-16CHARS',
      currency: 'EUR', computeVersion: 0,
    });
    await setDoc(doc(f, `${SL}/accounts/a1`), { name: 'Antiguo', order: 0 });
    await setDoc(doc(f, `${SL}/accounts/a1/tickets/t1`), {
      kind: 'manual', grandTotal: 100, paidByParticipantId: 'p1',
      spaceId: 'gr1',
    });
    await setDoc(doc(f, `${SL}/accounts/a1/tickets/t1/lines/l1`), {
      name: 'Café', totalPrice: 100,
      assignment: { type: 'unassigned', participants: {} },
    });
  });
});

describe('A11b: el grupo audita el ticket', () => {
  it('un miembro normal lee ticket, líneas, cuentas y participantes', async () => {
    const f = db(JORGE);
    await assertSucceeds(getDoc(doc(f, T)));
    await assertSucceeds(getDocs(collection(f, `${T}/lines`)));
    await assertSucceeds(getDocs(collection(f, `${SG}/accounts`)));
    await assertSucceeds(getDocs(collection(f, `${SG}/participants`)));
    // Y llega al ticket recorriendo las cuentas, que es lo que hace la app.
    await assertSucceeds(getDocs(collection(f, `${SG}/accounts/a1/tickets`)));
  });

  it('lo lee aunque no consuma nada de ese ticket', async () => {
    // l1 está asignada entera a p1 (Alba) y l2 a nadie: Jorge no aparece.
    const linea = await getDoc(doc(db(JORGE), `${T}/lines/l1`));
    if (linea.data().name !== 'Coca-Cola') {
      throw new Error('la línea no llegó completa al auditor');
    }
  });

  it('el propietario del grupo y un administrador también leen', async () => {
    for (const uid of [JEFA, ADMIN]) {
      await assertSucceeds(getDoc(doc(db(uid), T)));
      await assertSucceeds(getDocs(collection(db(uid), `${T}/lines`)));
    }
  });

  it('el dueño de la sesión conserva su lectura y su escritura', async () => {
    const f = db(ALBA);
    await assertSucceeds(getDoc(doc(f, T)));
    await assertSucceeds(updateDoc(doc(f, `${T}/lines/l1`), {
      name: 'Coca-Cola Zero',
    }));
  });

  // ── La otra mitad del contrato: leer NO es escribir ──────────────────
  // Ojo al alcance: desde A11c, quien ADMINISTRA el grupo sí corrige el
  // contenido del ticket (su autoridad y sus límites están en
  // group_ticket_correction.test.mjs). Lo que A11b fija es que auditar, por
  // sí solo, no escribe nada: eso se comprueba con el miembro normal.
  it('el auditor no toca una sola línea del ticket ajeno', async () => {
    const f = db(JORGE);
    await assertFails(updateDoc(doc(f, `${T}/lines/l1`), { totalPrice: 1 }));
    await assertFails(updateDoc(doc(f, `${T}/lines/l1`), { name: 'Otro' }));
    await assertFails(deleteDoc(doc(f, `${T}/lines/l1`)));
    await assertFails(setDoc(doc(f, `${T}/lines/nueva`), {
      name: 'Inventada', totalPrice: 100,
    }));
    // Y ni siquiera el administrador puede INVENTAR un producto: añadir
    // líneas sigue siendo del dueño de la sesión.
    for (const uid of [JEFA, ADMIN]) {
      await assertFails(setDoc(doc(db(uid), `${T}/lines/nueva`), {
        name: 'Inventada', totalPrice: 100,
      }));
    }
  });

  it('tampoco el ticket, la cuenta, el participante ni la sesión', async () => {
    const f = db(JORGE);
    await assertFails(updateDoc(doc(f, T), { grandTotal: 1 }));
    await assertFails(updateDoc(doc(f, T), { spaceId: 'gr2' }));
    await assertFails(deleteDoc(doc(f, T)));
    await assertFails(updateDoc(doc(f, `${SG}/accounts/a1`), { name: 'X' }));
    await assertFails(updateDoc(doc(f, `${SG}/participants/p1`), {
      claimedByDevice: JORGE,
    }));
    await assertFails(updateDoc(doc(f, SG), { status: 'closed' }));
    await assertFails(deleteDoc(doc(f, SG)));
  });

  // El documento de sesión guarda el `shareCode`: leerlo permitiría
  // fabricarse un guestAccess y con él EDITAR. Ampliar la lectura no puede
  // ampliar la escritura por la puerta de atrás.
  it('el grupo NO lee el documento de sesión (shareCode)', async () => {
    for (const uid of [JORGE, JEFA, ADMIN]) {
      await assertFails(getDoc(doc(db(uid), SG)));
    }
    // Por qué importa: el shareCode ES la credencial de invitado (ADR-012).
    // Quien lo conoce se da de alta como tal y con ello puede EDITAR líneas.
    // A11b no toca ese mecanismo; lo que hace es no regalar el secreto.
    await assertSucceeds(setDoc(doc(db(JORGE), `${SG}/guestAccess/${JORGE}`), {
      shareCode: 'SECRET-CODE-16CHARS',
    }));
  });

  it('la auditoría no arrastra liquidaciones ni actividad de la sesión', async () => {
    const f = db(JORGE);
    await assertFails(getDocs(collection(f, `${SG}/settlements`)));
    await assertFails(getDocs(collection(f, `${SG}/activity`)));
  });

  // ── Elegir MI consumo no es editar el ticket ─────────────────────────
  // Jorge (p2) está reclamado por su UID, igual que un invitado reclama su
  // nombre: la misma puerta de siempre, con la misma cerradura.
  it('el miembro marca y desmarca SU consumo en una línea del ticket', async () => {
    const f = db(JORGE);
    await assertSucceeds(updateDoc(doc(f, `${T}/lines/l2`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p2': true,
    }));
    await assertSucceeds(updateDoc(doc(f, `${T}/lines/l2`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p2': deleteField(),
    }));
  });

  it('no puede marcar por otro ni retirar lo que otro marcó', async () => {
    const f = db(JORGE);
    // Firmar como Alba: el pid no está reclamado por su UID.
    await assertFails(updateDoc(doc(f, `${T}/lines/l2`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p1',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p1': true,
    }));
    // Firmar como él mismo pero tocar la entrada de otro.
    await assertFails(updateDoc(doc(f, `${T}/lines/l1`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p1': deleteField(),
    }));
  });

  it('marcar consumo NO abre la puerta a editar el dato fuente', async () => {
    const f = db(JORGE);
    // Colar el precio en la misma escritura que la asignación.
    await assertFails(updateDoc(doc(f, `${T}/lines/l2`), {
      totalPrice: 1,
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p2': true,
    }));
    await assertFails(updateDoc(doc(f, `${T}/lines/l2`), { name: 'Caviar' }));
  });

  it('en un ticket a medias no se elige nada', async () => {
    await env.withSecurityRulesDisabled(async (ctx) => {
      await updateDoc(doc(ctx.firestore(), T), { splitModeOverride: 'equal' });
    });
    await assertFails(updateDoc(doc(db(JORGE), `${T}/lines/l2`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p2': true,
    }));
  });

  it('con la sesión cerrada tampoco', async () => {
    await env.withSecurityRulesDisabled(async (ctx) => {
      await updateDoc(doc(ctx.firestore(), SG), { status: 'closed' });
    });
    await assertFails(updateDoc(doc(db(JORGE), `${T}/lines/l2`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p2': true,
    }));
  });

  it('el propietario del grupo no elige por nadie: no es participante', async () => {
    // JEFA preside el grupo pero no está en este ticket. Administrar no es
    // consumir, y A11c no ha llegado.
    await assertFails(updateDoc(doc(db(JEFA), `${T}/lines/l2`), {
      'assignment.type': 'units',
      'assignment.schemaVersion': 2,
      'assignment.lastEditorPid': 'p2',
      'assignment.lastEditedUnit': 'u0',
      'assignment.units.u0.p2': true,
    }));
  });

  // ── Fuera del grupo no hay nada que ver ──────────────────────────────
  it('un miembro de OTRO grupo no lee nada', async () => {
    const f = db(AJENA);
    await assertFails(getDoc(doc(f, T)));
    await assertFails(getDocs(collection(f, `${T}/lines`)));
    await assertFails(getDocs(collection(f, `${SG}/participants`)));
    await assertFails(getDocs(collection(f, `${SG}/accounts`)));
  });

  it('un externo total tampoco, ni conociendo la ruta exacta', async () => {
    const f = db(EXTERNO);
    await assertFails(getDoc(doc(f, T)));
    await assertFails(getDoc(doc(f, `${T}/lines/l1`)));
    await assertFails(getDocs(collection(f, `${SG}/accounts`)));
    await assertFails(getDoc(doc(f, SG)));
  });

  it('ni el externo ni el de otro grupo pueden marcar consumo', async () => {
    for (const uid of [EXTERNO, AJENA]) {
      await assertFails(updateDoc(doc(db(uid), `${T}/lines/l2`), {
        'assignment.type': 'units',
        'assignment.schemaVersion': 2,
        'assignment.lastEditorPid': 'p2',
        'assignment.lastEditedUnit': 'u0',
        'assignment.units.u0.p2': true,
      }));
    }
  });

  // ── Lo que NO entra en la auditoría de grupo ─────────────────────────
  it('una RELACIÓN mantiene su comportamiento: no se audita como grupo', async () => {
    const f = db(PAREJA);
    await assertFails(getDoc(doc(f, `${SR}/accounts/a1/tickets/t1/lines/l1`)));
    await assertFails(getDocs(collection(f, `${SR}/participants`)));
    // El dueño de la sesión sigue leyendo la suya con normalidad.
    await assertSucceeds(
      getDoc(doc(db(JEFA), `${SR}/accounts/a1/tickets/t1/lines/l1`)));
  });

  it('una sesión SIN contexto no se abre por vincular un ticket a mano', async () => {
    const f = db(JORGE);
    // El RESUMEN del ticket vinculado sí se ve (P4), y eso no cambia.
    await assertSucceeds(getDoc(doc(f, `${SL}/accounts/a1/tickets/t1`)));
    // Pero la sesión legacy no se convierte en material del grupo.
    await assertFails(getDoc(doc(f, `${SL}/accounts/a1/tickets/t1/lines/l1`)));
    await assertFails(getDocs(collection(f, `${SL}/accounts`)));
  });

  it('quien deja de ser miembro deja de auditar', async () => {
    await assertSucceeds(getDoc(doc(db(JORGE), T)));
    await env.withSecurityRulesDisabled(async (ctx) => {
      await deleteDoc(doc(ctx.firestore(), `spaces/gr1/members/${JORGE}`));
    });
    await assertFails(getDoc(doc(db(JORGE), `${T}/lines/l1`)));
  });
});

File: apps/mobile/test/group_ticket_audit_test.dart
import 'package:fake_cloud_firestore/fake_cloud_firestore.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:salda_mobile/features/sessions/data/firestore_session_repository.dart';
import 'package:salda_mobile/features/sessions/domain/session_models.dart';
import 'package:salda_mobile/features/sessions/presentation/ticket_navigation.dart';
import 'package:salda_mobile/l10n/generated/app_localizations.dart';

import 'fakes.dart';

/// A11b: un miembro del grupo AUDITA el ticket de otro.
///
/// La autoridad la aplican Rules (backend/firestore/test/group_ticket_audit
/// y storage_receipt_access): `fake_cloud_firestore` no las evalúa. Lo que
/// se fija aquí es la otra mitad — que la pantalla muestre la evidencia
/// completa y NO ofrezca ni una acción que el auditor no pueda ejecutar.
///
/// El auditor se simula por donde de verdad se distingue: no puede leer
/// `sessions/{sid}` (ahí vive el shareCode), así que ese stream falla.
class _AuditorSessionRepository extends FirestoreSessionRepository {
  _AuditorSessionRepository({
    required super.firestore,
    required super.uid,
    required super.shareCodeFactory,
  });

  @override
  Stream<SessionDetail?> watchSession(String sessionId) => Stream.error(
    Exception('permission-denied: sessions/$sessionId'),
    StackTrace.current,
  );
}

const _ticketPath = 'sessions/s1/accounts/a1/tickets/t1';

Future<FakeFirebaseFirestore> _seed({String modo = 'byItem'}) async {
  final fake = FakeFirebaseFirestore();
  await fake.doc('spaces/gr1').set({
    'name': 'Piso',
    'ownerUid': 'uid-alba',
    'kind': 'group',
    'status': 'active',
    'schemaVersion': 2,
  });
  for (final uid in ['uid-alba', 'owner']) {
    await fake.doc('spaces/gr1/members/$uid').set({'uid': uid});
  }
  // La sesión la creó Alba dentro del grupo; quien mira es 'owner' (Jorge).
  await fake.doc('sessions/s1').set({
    'ownerUid': 'uid-alba',
    'kind': 'single',
    'name': 'Compra',
    'status': 'open',
    'splitModeDefault': modo,
    'shareCode': 'TEST-CODE-1234567890',
    'currency': 'EUR',
    'contextModelVersion': 1,
    'spaceId': 'gr1',
  });
  await fake.doc('sessions/s1/participants/p1').set({
    'name': 'Alba',
    'isOwner': true,
    'order': 0,
    'claimedByDevice': 'uid-alba',
  });
  await fake.doc('sessions/s1/participants/p2').set({
    'name': 'Jorge',
    'isOwner': false,
    'order': 1,
    'claimedByDevice': 'owner',
  });
  await fake.doc('sessions/s1/accounts/a1').set({'name': 'Súper', 'order': 0});
  await fake.doc(_ticketPath).set({
    'kind': 'manual',
    'grandTotal': 550,
    'paidByParticipantId': 'p1',
    'merchant': {'name': 'Familycash'},
    'spaceId': 'gr1',
    'contextModelVersion': 1,
    // Modo efectivo en el ticket: es lo único por lo que un miembro —que no
    // puede leer la sesión— sabe que el gasto se reparte por líneas.
    'splitModeOverride': modo,
  });
  // Consumida ENTERA por Alba: Jorge no participa y aun así debe verla.
  await fake.doc('$_ticketPath/lines/l1').set({
    'name': 'Coca-Cola',
    'order': 0,
    'quantityMilli': 1000,
    'unitPrice': 250,
    'totalPrice': 250,
    'unitIds': ['u0'],
    'assignment': {
      'type': 'units',
      'schemaVersion': 2,
      'units': {
        'u0': {'p1': true},
      },
    },
  });
  // Dos unidades sin reclamar: el creador las reparte con chips; el auditor
  // ve el mismo estado sin un solo control.
  await fake.doc('$_ticketPath/lines/l2').set({
    'name': 'Patatas',
    'order': 1,
    'quantityMilli': 2000,
    'unitPrice': 150,
    'totalPrice': 300,
    'unitIds': ['u0', 'u1'],
    'assignment': {
      'type': 'units',
      'schemaVersion': 2,
      'units': <String, Object?>{},
    },
  });
  return fake;
}

Future<void> _pump(
  WidgetTester tester,
  FakeFirebaseFirestore fake, {
  required bool auditor,
}) async {
  tester.view.physicalSize = const Size(800, 2000);
  tester.view.devicePixelRatio = 1.0;
  addTearDown(tester.view.reset);
  resetTicketNavigationDebounce();
  await tester.pumpWidget(
    ProviderScope(
      overrides: loggedInOverrides(
        firestore: fake,
        sessionRepository: auditor
            ? _AuditorSessionRepository(
                firestore: fake,
                uid: () => 'owner',
                shareCodeFactory: () => 'X',
              )
            : null,
      ),
      child: MaterialApp(
        localizationsDelegates: AppLocalizations.localizationsDelegates,
        supportedLocales: AppLocalizations.supportedLocales,
        // Por identificadores, como cualquier superficie del producto: el
        // auditor llega desde la lista de tickets del grupo, sin conocer la
        // cuenta en la que vive el ticket.
        home: const TicketRoute(sessionId: 's1', ticketId: 't1'),
      ),
    ),
  );
  await tester.pumpAndSettle();
}

void main() {
  testWidgets('el auditor ve el ticket entero: productos, precios y reparto', (
    tester,
  ) async {
    await _pump(tester, await _seed(), auditor: true);

    expect(find.text('Familycash'), findsWidgets);
    expect(find.text('Coca-Cola'), findsOneWidget);
    expect(find.text('Patatas'), findsOneWidget);
    // Precio por línea y total del ticket.
    expect(find.textContaining('2,50'), findsWidgets);
    expect(find.textContaining('3,00'), findsWidgets);
    expect(find.textContaining('5,50'), findsWidgets);
    // Y de quién es cada unidad: sin los nombres el reparto es ilegible.
    expect(find.textContaining('Alba'), findsWidgets);
  });

  testWidgets('no se le ofrece ninguna acción de EDICIÓN del ticket ajeno', (
    tester,
  ) async {
    await _pump(tester, await _seed(), auditor: true);

    // Compartir por enlace y vincular a un espacio son del dueño.
    expect(find.byIcon(Icons.link), findsNothing);
    expect(find.byIcon(Icons.group_work), findsNothing);
    expect(find.byIcon(Icons.group_work_outlined), findsNothing);
    // Migrar la línea al modelo de unidades reescribe la asignación entera:
    // es edición, no selección.
    expect(find.text('Repartir por unidades'), findsNothing);
  });

  // El núcleo de la corrección: leer y elegir son autoridades distintas.
  testWidgets('pero SÍ puede elegir su propio consumo, y lo firma como él', (
    tester,
  ) async {
    final fake = await _seed();
    await _pump(tester, fake, auditor: true);

    // Coca-Cola: una unidad, se marca tocando la fila.
    await tester.tap(find.text('Coca-Cola'));
    await tester.pumpAndSettle();
    // Patatas: dos unidades, se marca por chip.
    await tester.tap(find.widgetWithText(FilterChip, '1'));
    await tester.pumpAndSettle();

    final coca = await fake.doc('$_ticketPath/lines/l1').get();
    final patatas = await fake.doc('$_ticketPath/lines/l2').get();
    final cocaUnits =
        (coca.data()!['assignment'] as Map)['units'] as Map<String, dynamic>;
    final patatasAssignment = patatas.data()!['assignment'] as Map;

    // Marca SU unidad, con SU participante, sin tocar la de Alba.
    expect((cocaUnits['u0'] as Map)['p2'], true);
    expect((cocaUnits['u0'] as Map)['p1'], true);
    expect((coca.data()!['assignment'] as Map)['lastEditorPid'], 'p2');
    expect(
      ((patatasAssignment['units'] as Map)['u0'] as Map)['p2'],
      true,
    );
    expect(patatasAssignment['lastEditorPid'], 'p2');
    // Y nada del dato fuente se ha movido.
    expect(coca.data()!['name'], 'Coca-Cola');
    expect(coca.data()!['totalPrice'], 250);
  });

  // Reparto GLOBAL: aquí las líneas no deciden nada. Antes la pantalla
  // rotulaba cada producto como «sin reclamar (para Alba)» —la lectura del
  // otro modo— y quien auditaba concluía que no le tocaba nada.
  testWidgets('en un ticket a medias se explica el reparto y su parte', (
    tester,
  ) async {
    final fake = await _seed(modo: 'equal');
    // 15,96 € a medias entre Alba y Jorge.
    await fake.doc(_ticketPath).update({'grandTotal': 1596});
    await fake.doc('$_ticketPath/lines/l2').update({'totalPrice': 1346});
    await _pump(tester, fake, auditor: true);

    expect(
      find.textContaining('se reparte a partes iguales entre 2 personas'),
      findsOneWidget,
    );
    expect(find.textContaining('Te corresponden 7,98'), findsOneWidget);
    // Sigue auditando la compra: productos y precios originales.
    expect(find.text('Coca-Cola'), findsOneWidget);
    expect(find.text('Patatas'), findsOneWidget);
    expect(find.textContaining('2,50'), findsWidgets);
    expect(find.textContaining('15,96'), findsWidgets);
    // Sin controles de selección y sin rótulos del otro modo de reparto.
    expect(find.byType(FilterChip), findsNothing);
    expect(
      tester
          .widgetList<ListTile>(find.byType(ListTile))
          .every((tile) => tile.onTap == null),
      isTrue,
    );
    expect(find.textContaining('sin reclamar'), findsNothing);
  });

  testWidgets('sin participante propio no hay nada que elegir', (tester) async {
    final fake = await _seed();
    // Jorge está en el grupo, pero el anfitrión no lo metió en ESTE gasto.
    await fake.doc('sessions/s1/participants/p2').update({
      'claimedByDevice': 'uid-nadie',
    });
    await _pump(tester, fake, auditor: true);

    expect(find.text('Coca-Cola'), findsOneWidget); // lo sigue auditando
    expect(find.byType(FilterChip), findsNothing);
    expect(
      tester
          .widgetList<ListTile>(find.byType(ListTile))
          .every((tile) => tile.onTap == null),
      isTrue,
    );
  });

  testWidgets('el dueño de la sesión conserva sus acciones y su selección', (
    tester,
  ) async {
    final fake = await _seed();
    // Mismo ticket, pero ahora quien mira es su creador.
    await fake.doc('sessions/s1').update({'ownerUid': 'owner'});
    await fake.doc('sessions/s1/participants/p1').update({
      'name': 'Edgar',
      'claimedByDevice': 'owner',
    });
    await fake.doc('sessions/s1/participants/p2').update({
      'claimedByDevice': 'uid-jorge',
    });
    await _pump(tester, fake, auditor: false);

    expect(find.byIcon(Icons.link), findsOneWidget);
    expect(find.text('Coca-Cola'), findsOneWidget);
    // El creador sí elige lo que consumió (P2.1): chips en la línea de dos
    // unidades y fila pulsable en la de una.
    expect(
      tester
          .widgetList<FilterChip>(find.byType(FilterChip))
          .any((chip) => chip.onSelected != null),
      isTrue,
    );
    expect(
      tester
          .widgetList<ListTile>(find.byType(ListTile))
          .any((tile) => tile.onTap != null),
      isTrue,
    );
  });
}
